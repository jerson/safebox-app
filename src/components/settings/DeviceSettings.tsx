import React, { useState, useEffect, useRef } from 'react';
import {
  StyleSheet,
  ViewStyle,
  View,
  StyleProp,
  TextStyle,
  Platform
} from 'react-native';
import Colors from '../../modules/constants/Colors';
import SplitText from '../ui/SplitText';
import Button from '../ui/Button';
import Text from '../ui/Text';
import Biometrics from 'react-native-biometrics';
import Log from '../../modules/log/Log';
import useAnimatedState from '../hooks/useAnimatedState';
import AlertMessage from '../ui/AlertMessage';
import Client from '../../services/Client';
import { AddDeviceRequest, Device } from '../../proto/services_pb';
import DeviceInfo from 'react-native-device-info';
import Strings from '../../modules/format/Strings';
import NotAvailableDevice from '../help/NotAvailableDevice';
import List, { ListRef } from '../device/List';
import AppStorage from '../../modules/storage/AppStorage';
import SettingsStorage from '../../modules/storage/SettingsStorage';

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.grey1
  } as ViewStyle,
  content: {
    marginTop: 10,
    alignItems: 'center'
  } as ViewStyle,
  info: {} as ViewStyle,
  description: {
    fontSize: 12,
    color: Colors.grey5
  } as TextStyle,
  options: {
    alignItems: 'center'
  } as ViewStyle
});

export interface DeviceSettingsProps {
  style?: StyleProp<ViewStyle>;
}

const TAG = '[DeviceSettings]';
function DeviceSettings({ style }: DeviceSettingsProps) {
  const [biometricType, setBiometricType] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useAnimatedState('');

  const [isAddedCurrentDevice, setIsAddedCurrentDevice] = useAnimatedState(
    false
  );
  const listRef = useRef<ListRef>(null);

  const deviceUid = DeviceInfo.getUniqueID();

  useEffect(() => {
    checkBiometricType();
  }, []);

  const onLoadDevices = (devices: Device[]) => {
    const isAddedCurrentDevice = devices.some(device => {
      return device.getUid() === deviceUid;
    });
    setIsAddedCurrentDevice(isAddedCurrentDevice);
  };

  const checkBiometricType = async () => {
    try {
      const biometryType = await Biometrics.isSensorAvailable();

      if (biometryType === Biometrics.TouchID) {
        setBiometricType(Platform.OS === 'ios' ? 'TouchID' : 'Fingerprint');
      } else if (biometryType === Biometrics.FaceID) {
        setBiometricType(Platform.OS === 'ios' ? 'FaceID' : 'Face recognition');
      }
    } catch (e) {
      Log.warn(TAG, 'checkBiometricType', e);
    }
  };

  const savePublicKey = async (publicKey: string) => {
    const settings = await SettingsStorage.getFirst();
    AppStorage.write(() => {
      settings.biometricPublicKey = publicKey;
    });
  };

  const registerDevice = async (publicKey: string) => {
    setIsLoading(true);
    try {
      const request = new AddDeviceRequest();
      request.setName(DeviceInfo.getDeviceName() || '');
      request.setUid(deviceUid);
      request.setPublickey(publicKey);
      await Client.addDevice(request);
      savePublicKey(publicKey);
      listRef.current && listRef.current.load();
    } catch (e) {
      const message = Strings.getError(e);
      setError(message);
    }
    setIsLoading(false);
  };

  const addDevice = async () => {
    try {
      const publicKey = await Biometrics.createKeys('Please confirm');
      await registerDevice(publicKey);
    } catch (e) {
      const message = Strings.getError(e);
      setError(message);
    }
  };

  return (
    <View style={[styles.container, style]}>
      <SplitText title={'Biometric settings'} type={'Default'} />
      {!!error && <AlertMessage message={error} />}
      {!isAddedCurrentDevice && (
        <View style={styles.content}>
          <View style={styles.info}>
            <Text style={styles.description}>
              you can add additional authentication methods such as fingerprint
              or facial validation, depending on your device
            </Text>
          </View>
          {!biometricType && <NotAvailableDevice />}
          {!!biometricType && (
            <View style={styles.options}>
              <Button
                isLoading={isLoading}
                onPress={addDevice}
                title={`Add ${biometricType}`}
              />
            </View>
          )}
        </View>
      )}
      <List ref={listRef} currentDeviceUid={deviceUid} onLoad={onLoadDevices} />
    </View>
  );
}

export default DeviceSettings;
