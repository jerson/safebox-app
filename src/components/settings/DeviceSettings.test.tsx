import React from 'react';
import * as renderer from 'react-test-renderer';
import DeviceSettings, { DeviceSettingsProps } from './DeviceSettings';

const props: DeviceSettingsProps = {};
test('renders without crashing', () => {
  const snapshot = renderer.create(<DeviceSettings {...props} />).toJSON();
  expect(snapshot).toBeTruthy();
});

test('render just views', () => {
  const snapshot = renderer.create(<DeviceSettings {...props} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
