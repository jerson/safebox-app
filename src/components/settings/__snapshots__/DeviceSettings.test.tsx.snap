// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render just views 1`] = `
<_class
  style={
    Array [
      Object {
        "backgroundColor": "#fafdff",
      },
      undefined,
    ]
  }
>
  <_class
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "marginBottom": 5,
        },
      ]
    }
  >
    <_class
      style={
        Array [
          Object {
            "backgroundColor": "#9066f5",
            "height": 1,
            "left": 0,
            "opacity": 0.5,
            "position": "absolute",
            "right": 0,
            "top": 12,
          },
          Object {
            "backgroundColor": "#c2d0e7",
          },
        ]
      }
    />
    <_class
      style={
        Array [
          Object {
            "backgroundColor": "#4531cd",
            "borderRadius": 10,
            "paddingBottom": 4,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 6,
          },
          Object {
            "backgroundColor": "#6781a6",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Nunito",
              "fontSize": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {},
            Object {
              "color": "#FFFFFF",
              "fontSize": 13,
              "lineHeight": 14,
              "minWidth": 20,
              "textAlign": "center",
            },
          ]
        }
      >
        Biometric settings
      </Text>
    </_class>
  </_class>
  <_class
    style={
      Object {
        "alignItems": "center",
        "marginTop": 5,
      }
    }
  >
    <_class
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Nunito",
              "fontSize": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {},
            Object {
              "color": "#65748b",
              "fontSize": 12,
            },
          ]
        }
      >
        you can add additional authentication methods such as fingerprint or facial validation, depending on your device.
      </Text>
    </_class>
    <_class
      style={
        Object {
          "backgroundColor": "#f2f4f9",
          "borderRadius": 10,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Nunito",
              "fontSize": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {},
            Object {
              "color": "#38414e",
              "fontSize": 13,
            },
          ]
        }
      >
        your device does not have any supported method, try another device.
      </Text>
    </_class>
  </_class>
  <_class />
</_class>
`;
